Add ViewCommand and ViewExitCommand

brandon-nam opened the pull request:
Add ViewCommand and ViewExitCommand for enhancements of editing functionality.

brandon-nam committed:
Add ViewCommand and ViewExitCommand

Add ViewCommand and ViewExitCommand for enhancements of editing functionality.

brandon-nam committed:
Merge branch 'master' into edit-edit-command

brandon-nam committed:
Fix styling issues

Fix style issues to pass checkstyle test

brandon-nam committed:
Merge branch 'edit-edit-command' of https://github.com/brandon-nam/tp into edit-edit-command

brandon-nam committed:
Add JUnit tests for command executions in profile view page

While in profile view page, the command line should not allow execution of commands that are not listed in the ViewCommandParser.
Add JUnit tests to check if AddressBookParser and ViewCommandParser parse different commands in different views.

brandon-nam committed:
Fix styling issues

brandon-nam committed:
Fix styling issues

brandon-nam committed:
Add ViewCommandTest and ViewExitCommandTest

brandon-nam committed:
Add more JUnit Tests for ViewCommandTest and ViewModeParserTest

brandon-nam committed:
Fix styling issues

brandon-nam committed:
Add JUnit test for CommandResultTest

brandon-nam committed:
Fix styling issue

H1410101 commented on src/main/java/seedu/address/logic/LogicManager.java:
Interaction between logic manager and main window is currently kind of odd imo.
The following is what I understand from looking at the code:

Currently, `LogicManager` keeps track of `isInViewMode`, which 
(separation of concerns) seems like a UI thing to keep track of

Additionally, `MainWindow` does not explicitly keep track of view mode, and listens to CommandResult.
This relies on CommandResult implementing `isShowView()` and `isViewExit()` correctly, which seems unnecessary compared to the solution proposed below.

---

I propose there to be a source of truth between `LogicManager` and `MainWindow` (I personally prefer `MainWindow`, because it makes intuitive sense to me). This source of truth _explicitly_ tracks the current view mode. Additionally, `MainWindow` listens to _changes_ in the current view mode, comparing before the command is executed, and after the command is executed.

This also avoids potential bugs where `MainWindow` de-synchronizes with `LogicManager` (such as the `LogicManager`'s `isInViewMode` being `false` while the person's profile is being displayed).

brandon-nam replied:
I understand your proposal and it makes sense. 

I found out that with StackPane.isVisible() I can handle the visibility boolean check in the MainWindow. This implements your suggestion regarding MainWindow keeping track of the visibility status. 

To utilize StackPane.isVisible(), I created a new StackPane that is similar to the already existing personListPlaceholder StackPane in my latest commit. With isVisible() boolean check and StackPane.setVisible(), the system toggles the visibility between the two StackPanes.

H1410101 commented on src/main/java/seedu/address/ui/PersonProfile.java:
```diff
@@ -0,0 +1,57 @@
+        phone.setText(person.getPhone().value);
+        address.setText(person.getAddress().value);
+        email.setText(person.getEmail().value);
+        person.getTags().stream()
```
Currently throws error for me.

Steps to reproduce:
1. Start application.
2. Write `view 2`, press enter.

Result:
- `person.toString` printed to UI, as expected
- PersonProfile does not appear; unexpected

Call chain:
- `handleCommandEntered` @ CommandBox:45
- (lambda)
- `executeCommand` @ `MainWindow`:207
- `handleView` @ `MainWindow`:176
- `<init>` @ `PersonProfile`:53

brandon-nam replied:
The issue was with the data representation, since the person's tag was not properly being displayed in the PersonProfile.fxml implementation in this commit. 

The later commits properly display a person's tags, although further Ui styling must be done soon.

H1410101 reviewed (dismissed):
Main concerns:
- possible bug introduced due to lack of single source of truth for view mode between `MainWindow` and `LogicManager`
- unexpected behaviour on `view` command

brandon-nam committed:
Fix data representation error and add personProfilePlaceholder StackPane

Currently the PersonProfile FXML has issues displaying tags which causes the error.
Fix the tag display to prevent the error.
Using the already existing setVisible() method of StackPane can toggle the visibility of the fosterer list and the profile view.
Add a personProfilePlaceholder StackPane below the personListPlaceholder StackPane and allow the visibility to toggle between them.

brandon-nam committed:
Add JUnit tests for getIsViewCommand() and getIsViewExitCommand()

nananakx-x commented on src/main/java/seedu/address/logic/parser/AddCommandParser.java:
```diff
@@ -22,7 +22,7 @@
  * Parses input arguments and creates a new AddCommand object
  */
-public class AddCommandParser implements Parser<AddCommand> {
+public class AddCommandParser implements CommandParser<AddCommand> {
```
Wait why did you do this haha I didn't change this part for addcommandparser and I feel like if its changed to commandparser the add function might not work properly :(

nananakx-x reviewed (dismissed):
LGTM! PLS PLS PLS PLS PLS just make sure add command still works as intended, all the tests can pass, delete multiple also work!! Refer to Page 10 onwatds of this for add command test cases, https://docs.google.com/document/d/1TIactGBiN45u30Nv6yOLV3QEPRaXFWWGHKvTEaHOsk0/edit#heading=h.hhevn0icya3z

brandon-nam committed:
Merge branch 'master' into edit-edit-command

nananakx-x reviewed (approved):
Refer to previous approving comment LOL
